File: src/main/java/com/example/hypocaust/db/BaseEntity.java
package com.example.hypocaust.db;

import com.github.f4b6a3.uuid.UuidCreator;
import jakarta.persistence.Column;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import java.time.Instant;
import java.util.UUID;
import lombok.Getter;

@Getter
@MappedSuperclass
public abstract class BaseEntity {

  @Id
  private UUID id;

  @Column(nullable = false, updatable = false)
  private Instant createdAt;

  protected BaseEntity() {
    this.id = UuidCreator.getTimeOrderedEpoch();
    this.createdAt = Instant.now();
  }

}

File: src/main/kotlin/com/example/the_machine/domain/BaseEntity.kt
package com.example.the_machine.domain

import com.example.the_machine.common.UuidV7
import jakarta.persistence.Column
import jakarta.persistence.Id
import jakarta.persistence.MappedSuperclass
import org.hibernate.annotations.CreationTimestamp
import java.time.Instant
import java.util.*

@MappedSuperclass
abstract class BaseEntity {

  @Id
  val id: UUID = UuidV7.newId()

  @Column(nullable = false, updatable = false)
  @CreationTimestamp
  val createdAt: Instant? = null

}